core/src/ldml/ldml_markers.cpp

```diff
@@ -56,52 +57,44 @@ bool normalize_nfd(std::u16string &str) {
   return normalize(nfd, str, status);
 }
 
-bool normalize_nfd_markers_segment(std::u16string &str, marker_map &map, marker_encoding encoding) {
-  std::u32string rstr = km::core::kmx::u16string_to_u32string(str);
-  if(!normalize_nfd_markers_segment(rstr, map, encoding)) {
-    return false;
-  } else {
-    str = km::core::kmx::u32string_to_u16string(rstr);
-    return true;
-  }
-}
-
-static void add_back_markers(std::u32string &str, const std::u32string &src, const marker_map &map, marker_encoding encoding) {
+static void add_back_markers(std::u32string &str, const std::u32string &src, marker_map &map, marker_encoding encoding) {
   // need to reconstitute.
   marker_map map2(map);  // make a copy of the map
   // clear the string
   str.clear();
-  // add the end-of-text marker
-  {
-    const auto ch = MARKER_BEFORE_EOT;
-    const auto m  = map2.find(ch);
-    if (m != map2.end()) {
-      for (auto q = (m->second).rbegin(); q < (m->second).rend(); q++) {
-        prepend_marker(str, *q, encoding);
-      }
-      map2.erase(ch);  // remove it
-    }
+  // iterator over the marker map
+  auto marki = map2.rbegin();
+
+  // add any end-of-text markers
+  while(marki != map2.rend() && marki->first == MARKER_BEFORE_EOT) {
+    prepend_marker(str, (marki++)->second, encoding);
   }
+
   // go from end to beginning of string
   for (auto p = src.rbegin(); p != src.rend(); p++) {
     const auto ch = *p;
     str.insert(0, 1, ch);  // prepend
 
-    const auto m = map2.find(ch);
-    if (m != map2.end()) {
-      for (auto q = (m->second).rbegin(); q < (m->second).rend(); q++) {
-        prepend_marker(str, *q, encoding);
+    // add the markers at the end of the list first.
+    for (; marki != map2.rend() && marki->first == ch; marki++) {
+      if (marki->second != 0) {
+        // set to '0' if already applied
+        prepend_marker(str, marki->second, encoding);
+        marki->second = 0; // mark as already applied
+      }
+    }
+
+    // now, add any out of order markers.
+    for (auto marki2 = marki; marki2 != map2.rend(); marki2++) {
+      if (marki2->second != 0 && marki2->first == ch) {
+        prepend_marker(str, marki2->second, encoding);
+        marki2->second = 0; // mark as already applied
       }
-      map2.erase(ch);  // remove it
     }
   }
+//  assert(marki == map.rend()); // that we consumed everything
```

Comment:
Commented code here?

Reply:
good catch Should have been map2, but it wasn't the right assert. added a better assert now.